Implement Sync Todoist -> Obsidian

Heiss opened the pull request:
Hi,

because i need this functionality for my own workflow, i implement it myself, today.
This is my first contribution to obsidian, so feel free to give some hints. Also there were some mistakes by myself with my IDE, because it tries to force me to use the code linter. But i do not found any information about the settings, the author used.

After this PR you can now:
- create a task in Todoist and it gets synced with obsidian
  - new setting options added for two modes
  - the daily note mode adds the tasks to the daily note
  - the template note mode adds the tasks to a separate note or a note specific to corresponding project
  - both modes supports `append to file` and `Insert After` a specified text
- you can add a task to obsidian as before and if something goes wrong, it fixes now automatically itself
  - it searches the vault for tasks with the same task text from todoist
  - this will helps to get a correct database state, if there was a network / code error
- fixes some minor typos and bugs i found on the go, e.g. the settings shows the selected default project twice

This should implement the most of wanted features for #53 

Missing things:
- Projects notes will not be deleted in obsidian, when deleted in Todoist
- tags are not supported right now, so they are missing in obsidian, when added in Todoist

So hopefully, we can get a step closer to the ultimate sync plugin.

One thing, i would like to add: Maybe we should start to translate everything to english, so more people can help. Also we should introduce a common linter config. I have no hard feelings here, so feel free to decide. :)

I am looking forward to hear from you.
Have a great day.

Heiss committed:
add Settings for Pull from Todoist

Heiss committed:
respect new pullProjectId setting in filter.

Heiss committed:
add file operation for daily mode

Heiss committed:
fix bugs in settings

Heiss committed:
fixed a bug in default project selection

Heiss committed:
fixed a bug in default project selection

Heiss committed:
fixes git error

Heiss committed:
add links to settings

Heiss committed:
implement template mode

Heiss committed:
fixes a typo

Heiss committed:
fixes automatically tasks, which throw errors, because unbound to todoist

Heiss committed:
template respects title and date now

Heiss committed:
remove unused console logs

HeroBlackInk commented:
This messy code, written a few months ago, is difficult for me to read, and there are also numerous Chinese comments interspersed throughout... However, you managed to implement a new feature in just two days. I am truly amazed... Thumbs up!

HeroBlackInk commented:
Thank you very much for your contribution. I need some time to review the pull request, please be patient.

Heiss commented:
You are welcome. 👍

HeroBlackInk commented on src/syncModule.ts:
```diff
@@ -758,63 +758,83 @@ export class TodoistSync  {
             // 找出 task id 存在于 Obsidian 中的 task activity
             const result2 = result1.filter(
-            (objA) => savedTasks.some((objB) => objB.id === objA.object_id)
+            (objA) => !savedTasks.some((objB) => objB.id === objA.object_id)
```
Due to the fact that many tasks in Todoist were not created in Obsidian and do not require synchronization, the variable **result2** has been modified to remove events  that do not require further processing. No modifications are necessary in this line.
```suggestion
            (objA) => savedTasks.some((objB) => objB.id === objA.object_id)
```

HeroBlackInk commented on src/syncModule.ts:
```diff
@@ -758,63 +758,83 @@ export class TodoistSync  {
             // 找出 task id 存在于 Obsidian 中的 note activity
             const result3 = result1.filter(
-                (objA) => savedTasks.some((objB) => objB.id === objA.parent_item_id)
+                (objA) => !savedTasks.some((objB) => objB.id === objA.parent_item_id)
```
In the events object that related to comments/notes, the parent_item_id corresponds to the task_id. Since many tasks in Todoist were not created in Obsidian and do not require synchronization, I have removed events that do not need further processing and saved them in the variable **result3**. No modifications are required on this line.
```suggestion
            (objA) => savedTasks.some((objB) => objB.id === objA.parent_item_id)
```

HeroBlackInk replied:
In the events object that related to comments/notes, the parent_item_id corresponds to the task_id. Since many tasks in Todoist were not created in Obsidian and do not require synchronization, I have removed events that do not need further processing and saved them in the variable **result3**. No modifications are required on this line.
```suggestion
                (objA) => savedTasks.some((objB) => objB.id === objA.parent_item_id)
```

HeroBlackInk commented on src/syncModule.ts:
```diff
@@ -758,63 +758,83 @@ export class TodoistSync  {
-    
-    
+
+            console.log(result1)
```
I use a new variable result4 to save the new task events that need to be processed.
```suggestion
             const result4 = result1.filter(
            (objA) => !savedTasks.some((objB) => objB.id === objA.object_id)
            )
```

HeroBlackInk commented on src/syncModule.ts:
```diff
@@ -758,63 +758,83 @@ export class TodoistSync  {
 
             //Items updated (only changes to content, description, due_date and responsible_uid)
             const unsynchronized_item_updated_events = this.plugin.todoistSyncAPI.filterActivityEvents(result2, { event_type: 'updated', object_type: 'item' })
+            const unsynchronized_item_added_events = this.plugin.todoistSyncAPI.filterActivityEvents(result2, { event_type: 'added', object_type: 'item' })
```
Extract new tasks events that need to be synchronized to obsidian from the result4 variable.
```suggestion
            const unsynchronized_item_added_events = this.plugin.todoistSyncAPI.filterActivityEvents(result4, { event_type: 'added', object_type: 'item' })
```

HeroBlackInk commented on src/todoistSyncAPI.ts:
```diff
@@ -165,7 +165,7 @@ export class TodoistSyncAPI   {
         const allActivityEvents = allActivity.events
         //client中不包含obsidian 的activity
-        const filteredArray = allActivityEvents.filter(obj => !obj.extra_data.client?.includes("obsidian")); 
+        const filteredArray = allActivityEvents.filter(obj => !obj.extra_data.client?.includes("obsidian") && this.plugin.settings.pullFromProjectId.includes(obj.parent_project_id) );
```
If the judgment conditions in the latter part are added, it will result in some tasks created in Obsidian that cannot be synchronized back to Obsidian after being modified in Todoist.
```suggestion
        const filteredArray = allActivityEvents.filter(obj => !obj.extra_data.client?.includes("obsidian")
```

Heiss replied:
Do i understand this correctly? This condition is not needed, if we implement a result4 object to select which tasks to create?

HeroBlackInk reviewed (changes requested):
Thank you for your contribution. I have thoroughly reviewed the code, and there are a few areas that require some modifications.

Heiss committed:
Update src/todoistSyncAPI.ts

If the judgment conditions in the latter part are added, it will result in some tasks created in Obsidian that cannot be synchronized back to Obsidian after being modified in Todoist.

Co-authored-by: HeroBlackInk <drv1@outlook.com>

Heiss committed:
Apply suggestions from code review

Add a result4 object to track "tasks added" event. With this changes, it does not need the previous "dirty fixes", which adds bugs and flows

Co-authored-by: HeroBlackInk <drv1@outlook.com>

Heiss committed:
fixes missing brace from github reviews

Heiss committed:
add a new option to select, if you want to strip tags from the text or not.

Heiss committed:
now settings checks, if daily notes are active. Otherwise does not append this option.

Heiss committed:
add comment for result4

Heiss committed:
add a check for sync for daily note plugin and fixes an error with createDailyNote

Heiss committed:
adds labels as tags from todoist

Heiss committed:
fixes typo in settings and fileoperation

Heiss committed:
move the tags behind #todoist

Heiss committed:
move the setting for removeTagsWithText between syncs and remove #todoist regardles of this setting.

Heiss committed:
move auto generated tags behind #todoist

Heiss commented:
I added the labels / tags to new tasks coming from todoist. In the same situation, i try to implement a sync, but I don't find a way to implement with the REST api... because todoist do not show an event for this kind of change.

HeroBlackInk commented:
> I added the labels / tags to new tasks coming from todoist. In the same situation, i try to implement a sync, but I don't find a way to implement with the REST api... because todoist do not show an event for this kind of change.

Yes, the information provided by the sync API is more comprehensive, but I haven't finished this part yet.

Heiss committed:
add sync for updated labels

this triggers a full backup sync to get all tasks. there should be a better solution (without webhook) for this, but i cannot find it.

Heiss committed:
add text to console output

Heiss committed:
add text to console output and try to implement labels update

Heiss committed:
finish work with sync for labels / tags

Heiss committed:
add a check for labels to handle an error

Heiss committed:
add setting check to disable tag sync by default.

Because it is not able to handle a situation where todoist and obsidian where changed.

Heiss commented:
The labels will now be synced with the help of full sync function. Currently, it checks if there are labels in tasks (or are missing) and places the tags at the end of line or remove them.

This assumes, that the user does not change anything in the vault by hand (e.g. with obsidian) because they are not send to todoist. If there are changes in todoist, too, we get a race condition, which is undefinied behaviour.
So i add a check in settings to disable this feature per default.

Maybe you know a better workflow. But i cannot say, which change comes earlier in time, so we have here a first come first serve situation. :(

HeroBlackInk commented:
I would like to invite you to become a collaborator and have also created a dev branch where you can submit your contributions without requiring approval. This will allow me to test and modify the code more conveniently.

HeroBlackInk commented:
Due to rate limits, each user is allowed to make a maximum of 45 full sync requests within a 15-minute period. It is recommended to utilize the REST API's '**Get active tasks**' feature, which allows a maximum of 450 REST API requests per user within the same 15-minute timeframe.
```
import { TodoistApi } from "@doist/todoist-api-typescript"

const api = new TodoistApi("0123456789abcdef0123456789")

api.getTasks()
    .then((tasks) => console.log(tasks))
    .catch((error) => console.log(error))

```

Example response:

> [
>     {
>         creatorId: "2671355",
>         createdAt: "2019-12-11T22:36:50.000000Z",
>         assigneeId: "2671362",
>         assignerId: "2671355",
>         commentCount: 10,
>         isCompleted: false,
>         content: "Buy Milk",
>         description: "",
>         due: {
>             date: "2016-09-01",
>             isRecurring: false,
>             datetime: "2016-09-01T09:00:00.000000Z",
>             string: "tomorrow at 12",
>             timezone: "Europe/Moscow"
>         },
>         "duration": null,
>         id: "2995104339",
>         labels: ["Food", "Shopping"],
>         order: 1,
>         priority: 1,
>         projectId: "2203306141",
>         sectionId: "7025",
>         parentId: "2995104589",
>         url: "https://todoist.com/showTask?id=2995104339"
>     }
> ]

Returns a JSON-encoded array containing all active tasks.

HeroBlackInk commented:
Thank you very much for your contribution.Please withdraw this merge request and merge it into the dev branch for testing convenience.

Heiss commented:
Done. Changes are now pushed in dev branch. PR #67

mgruca commented:
Hi @HeroBlackInk , are you planning a release that includes this feature?
